pilot/pkg/xds/workloadentry.go

```diff
@@ -0,0 +1,219 @@
+	}
+	if err := sg.Store.Delete(gvk.WorkloadEntry, wle.Name, wle.Namespace); err != nil {
+		adsLog.Warnf("failed cleaning up auto-registered WorkloadEntry %s/%s: %v", wle.Namespace, wle.Name, err)
+	}
+}
+
+func shouldCleanupEntry(wle config.Config) bool {
+	// don't clean-up if connected or non-autoregistered WorkloadEntries
+	_, ok := wle.Annotations[WorkloadControllerAnnotation]
+	if wle.Annotations[AutoRegistrationGroupAnnotation] == "" || ok {
+		return false
+	}
+
+	disconnUnixTime, err := strconv.ParseInt(wle.Annotations[DisconnectedAtAnnotation], 10, 64)
+	if err != nil {
+		// remove workload entries with invalid disconnect times - they need to be re-registered and fixed.
+		adsLog.Warnf("invalid disconnect time for WorkloadEntry %s/%s: %s", wle.Annotations[DisconnectedAtAnnotation])
+	}
+	disconnAt := time.Unix(0, disconnUnixTime)
+	// if we haven't passed the grace period, don't cleanup
+	if err == nil && time.Since(disconnAt) < features.WorkloadEntryCleanupGracePeriod {
+		return false
+	}
+
+	return true
+}
+
+func setConnectMeta(c *config.Config, controller string, con *Connection) {
+	c.Annotations[WorkloadControllerAnnotation] = controller
+	c.Annotations[ConnectedAtAnnotation] = strconv.FormatInt(con.Connect.UnixNano(), 10)
+}
+
+func workloadEntryFromGroup(name string, proxy *model.Proxy, groupCfg *config.Config) *config.Config {
+	group := groupCfg.Spec.(*v1alpha3.WorkloadGroup)
+	entry := group.Template.DeepCopy()
+	entry.Address = proxy.IPAddresses[0]
+	if proxy.Metadata.Network != "" {
+		entry.Network = proxy.Metadata.Network
+	}
+	if proxy.Locality != nil {
+		entry.Locality = util.LocalityToString(proxy.Locality)
+	}
+	return &config.Config{
+		Meta: config.Meta{
+			GroupVersionKind: gvk.WorkloadEntry,
+			Name:             name,
+			Namespace:        proxy.Metadata.Namespace,
+			Labels:           mergeLabels(entry.Labels, proxy.Metadata.Labels),
+			Annotations:      map[string]string{AutoRegistrationGroupAnnotation: groupCfg.Name},
+		},
+		Spec: entry,
+		// TODO status fields used for garbage collection
+		Status: nil,
+	}
+}
+
+func mergeLabels(labels ...map[string]string) map[string]string {
+	if len(labels) == 0 {
+		return map[string]string{}
+	}
+	out := make(map[string]string, len(labels)*len(labels[1]))
+	for _, lm := range labels {
+		for k, v := range lm {
+			out[k] = v
+		}
+	}
+	return out
+}
+
+func autoregisteredWorkloadEntryName(proxy *model.Proxy) string {
+	if proxy.Metadata.AutoRegisterGroup == "" {
+		return ""
+	}
+	if len(proxy.IPAddresses) == 0 {
+		adsLog.Errorf("auto registration of %v failed: missing IP addresses", proxy.ID)
+		return ""
+	}
+	if len(proxy.Metadata.Namespace) == 0 {
+		adsLog.Errorf("auto registration of %v failed: missing namespace", proxy.ID)
+		return ""
+	}
+	p := []string{proxy.Metadata.AutoRegisterGroup, proxy.IPAddresses[0]}
+	if proxy.Metadata.Network != "" {
+		p = append(p, proxy.Metadata.Network)
+	}
+	return strings.Join(p, "-")
```

Comment:
we may need to add some validation for max len. Not sure what we will do if it is exceeded though...

Reply:
For now, I'm cutting it from the front (to preserve the more unique parts) and logging a warning. We should probably document this and maybe even add validation to WorkloadGroup creation to limit the name length. `network` name could still cause us to exceed this limit though.... 

I think it's probably better to just leave the too-long name and fail, so we don't hide this issue from users.